src/Microsoft.OData.Core/ODataRawInputContext.cs

```diff
@@ -251,5 +228,146 @@ private object ReadRawValue(IEdmPrimitiveTypeReference expectedPrimitiveTypeRefe
 
             return rawValue;
         }
+
+        /// <summary>
+        /// Asynchronously read a top-level value.
+        /// </summary>
+        /// <param name="expectedPrimitiveTypeReference">The expected primitive type for the value to be read; null if no expected type is available.</param>
+        /// <returns>
+        /// A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains an object representing the read value.
+        /// </returns>
+        private async Task<object> ReadValueImplementationAsync(IEdmPrimitiveTypeReference expectedPrimitiveTypeReference)
+        {
+            object value;
+
+            // If an expected primitive type is specified it trumps the content type/reader payload kind
+            bool readBinary = ReadAsBinary(expectedPrimitiveTypeReference);
+
+            if (readBinary)
+            {
+                value = await this.ReadBinaryValueAsync()
+                    .ConfigureAwait(false);
+            }
+            else
+            {
+                Debug.Assert(this.textReader == null, "this.textReader == null");
+                this.InitializeTextReader();
+                value = await this.ReadRawValueAsync(expectedPrimitiveTypeReference)
+                    .ConfigureAwait(false);
+            }
+
+            return value;
+        }
+
+        /// <summary>
+        /// Asynchronously read the binary value from the stream.
+        /// </summary>
+        /// <returns>
+        /// A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains a byte array containing all the data read.
+        /// </returns>
+        private async Task<byte[]> ReadBinaryValueAsync()
+        {
+            byte[] data;
+
+            long numberOfBytesRead = 0;
+            int result;
+            List<byte[]> byteData = new List<byte[]>();
+
+            do
+            {
+                data = new byte[BufferSize];
+                result = await this.stream.ReadAsync(data, 0, data.Length)
+                    .ConfigureAwait(false);
+                numberOfBytesRead += result;
+                byteData.Add(data);
+            }
+            while (result == data.Length);

```

Comment:
suggestion: Maybe a small improvement would be for seekable streams check `CanSeek` Then read to the stream end as an small optimization for seekable streams.

Reply:
According to one of the comments above, all the data has already been buffered and the reading is probably in memory. If that's the case, it's like to be a seekable stream.

Reply:
@habbes See my comment [here](https://github.com/OData/odata.net/pull/2364/files#r844296036)

Reply:
@marabooy I implemented your suggestion in both `ReadBinaryValue` and `ReadBinaryValueAsync`

Reply:
@habbes @marabooy Added tests for reading from non-seekable stream